File: 011-数值的整数次方/power.py
# -*- coding:utf-8 -*-
class Solution:
    def Power(self, base, exponent):

        ex=exponent
        if exponent<0:
            ex=-exponent
        ans=self.ksm(base,ex)
        if exponent<0:
            return 1/ans
        return ans
    def ksm(self,base,exponent):
        ans=1
        while(exponent):
            if exponent & 1==1:
                ans=ans*base
            base=base*base
            exponent>>=1
        return ans

File: 011-数值的整数次方/power.cpp
class Solution {
public:
    double Power(double base, int exponent) {
        if (exponent==0){
            return 1;
        }
        else if (exponent==1){
            return base;
        }
        int e=exponent;
        if (exponent<0){
            e=-e;
        }
        double ans=Power(base,e>>1);
        ans=ans*ans;
        if ((e & 1)==1)
        {
            ans*=base;
        }
        if (exponent>0){
            return ans;
        }
        else{
            return 1/ans;
        }
    }
};
